lib/galaxy/jobs/runners/kubernetes.py

```diff
@@ -430,6 +425,38 @@ def check_watched_item(self, job_state):
             self.mark_as_failed(job_state)
             return job_state
 
+    def _handle_job_failure(self, job, job_state, reason=None):
+        self.__produce_log_file(job_state)
+        error_file = open(job_state.error_file, 'w')
+        if reason == "OOM":
+            error_file.write("Job killed after running out of memory. Try with more memory.\n")
+            job_state.fail_message = "Tool failed due to insufficient memory. Try with more memory."
+            job_state.runner_state = JobState.runner_states.MEMORY_LIMIT_REACHED
+        else:
+            error_file.write("Exceeded max number of Kubernetes pod retrials allowed for job\n")
+            job_state.fail_message = "More pods failed than allowed. See stdout for pods details."
+        error_file.close()
+        job_state.running = False
+        self.mark_as_failed(job_state)
```

Comment:
I'm seeing that none of the other runners call this directly anymore (except the chronos one that I think was based on this one here). So you should probably not be calling this directly - I think you want to make sure you can call ``self.fail_job()`` and let it handle those details. It is what triggers the resubmission. Can you rework that to happen?

Reply:
Sure! Will probably get back to this next week. Thanks for the review John!

Reply:
Thanks for this input John! I'm looking at the other runners, but I'm struggling a bit as they seem to be doing different things:

* condor: on the check_watched_items, is doing:

```
cjs.fail_message = "Cluster could not complete job"
self.work_queue.put((self.fail_job, cjs))
continue
```
but under a difference case doing, it does set the `cjs.failed` var but uses `self.finish_job` instead:
```
if job_failed:
     log.debug("(%s/%s) job failed" % (galaxy_id_tag, job_id))
     cjs.failed = True
     self.work_queue.put((self.finish_job, cjs))
```

* GoDocker: this one resembles the k8s one, takes care of creating a log file and then using the `self.marked_as_failed(job_state)` approach

* PBS: it overrides `fail_job` and does something like:
```
pbs_job_state.job_wrapper.fail(pbs_job_state.fail_message)
```

* slurm: 
```
if drmaa_state == self.drmaa_job_states.FAILED:
       ajs.fail_message += '\nPlease click the bug icon to report this problem if you need help.'
       ajs.stop_job = False
       self.work_queue.put((self.fail_job, ajs))
       return
```

* drmaa: uses fail_job, but doesn't set `ajs.*` variables
```
elif self.runner_params[state_param] == model.Job.states.ERROR:
       log.warning("(%s/%s) job will now be errored", galaxy_id_tag, external_job_id)
       self.work_queue.put((self.fail_job, ajs))
```

So I'm a bit confused, which is the correct construct to use? Is it:

```
ajs.failed = True
self.work_queue.put(self.fail_job, ajs)
```

To add to this confusion, the figure at these [Galaxy runner guidelines](https://docs.galaxyproject.org/en/master/dev/build_a_job_runner.html) which I recently came accross state in figure for Stage 3 to use `mark_as_failed` method inside check_watched_item. Could you please clarify and point to desired examples please? Thanks!

Reply:
Fair enough - if this part is working for you we can keep it as is. We need to work toward figuring out a best practice and get that resubmission logic used everywhere but the state of things is clearly all over the map. Thanks for looking into this though - it is nice we are at least catching the reason and such - this is really nice!

Reply:
One detail though, I would like to have the resubmission working. You said that there is a particular something that needs to be invoked for that, how should be done? Currently the resubmission is not working.

Reply:
Well, I guess we can leave that for later as well for another PR.

Reply:
Digging into it a bit more - it does look like it should work to me the way you are doing it.

I'd put some logging in around this line in ``lib/galaxy/jobs/runners/__init__.py`` - make sure it is getting there and that job_state.runner_state is correct. 
```
self._handle_runner_state('failure', job_state)
```

I'd also try with completely static job destinations - if the problem is that resubmission isn't working because of dynamic job destinations - we should create an issue and address that. There are some ideas of potential problems and how to address them in on this PR from last week (https://github.com/galaxyproject/galaxy/pull/5139).

Regardless - once we merge this as is - can you create an issue so we can track this further?